### PR TITLE
Convolution on habana accelerators

### PR DESCRIPTION
**Pull request by jerome-habana**
## What does this PR do?

 Gaudi HW performs convolution operations with filter (weights) in filters last format - RSCK format where:
 R = height of the filter ,S = width of the filter, C = number of channels per filter,K = number of filters
 The default PyTorch convolution weight ordering is ‘filters first’ (KCRS).
 Therefore a re-ordering/permutation of all the convolution weights from KCRS to RSCK format is required
 before convolution operations

Added a new helper to aid in permutes before moving the model to Gaudi

Fixes #\<issue_number>

## Before submitting

- [ ] Was this **discussed/approved** via a GitHub issue? (not for typos and docs)
- [ ] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [ ] Did you make sure your **PR does only one thing**, instead of bundling different changes together?
- [ ] Did you make sure to **update the documentation** with your changes? (if necessary)
- [ ] Did you write any **new necessary tests**? (not for typos and docs)
- [ ] Did you verify new and **existing tests pass** locally with your changes?
- [ ] Did you list all the **breaking changes** introduced by this pull request?
- [ ] Did you **update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)**? (not for typos, docs, test updates, or minor internal changes/refactors)

## PR review

Anyone in the community is welcome to review the PR.
Before you start reviewing, make sure you have read the [review guidelines](https://github.com/PyTorchLightning/pytorch-lightning/wiki/Review-guidelines). In short, see the following bullet-list:

- [ ] Is this pull request ready for review? (if not, please submit in draft mode)
- [ ] Check that all items from **Before submitting** are resolved
- [ ] Make sure the title is self-explanatory and the description concisely explains the PR
- [ ] Add labels and milestones (and optionally projects) to the PR so it can be classified

## Did you have fun?

Make sure you had fun coding 🙃

### CODE REVIEW

File: docs/source/accelerators/hpu.rst [hunk 1]
```diff
@@ -112,6 +112,50 @@ This enables advanced users to provide their own BF16 and FP32 operator list ins
 
 For more details, please refer to `PyTorch Mixed Precision Training on Gaudi <https://docs.habana.ai/en/latest/PyTorch_User_Guide/PyTorch_User_Guide.html#pytorch-mixed-precision-training-on-gaudi>`__.
 
+
+Enabling Models with Convolution layers
+---------------------------------------
+Convolution operations are central to vision topologies like ResNet. Gaudi HW performs convolution operations with filter (weights) in filters last format - RSCK format where:
+R = height of the filter; S = width of the filter; C = number of channels per filter; K = number of filters
```
Comment (justusschock) [inline docs/source/accelerators/hpu.rst:119]: Can we use the terms PyTorch uses in its docs for this: KCHW?

File: pl_examples/hpu_examples/convolution_example.py [hunk 2]
```diff
@@ -0,0 +1,93 @@
...
+val_loader = DataLoader(val_ds, batch_size=16)
+
+# Initialize a trainer
+trainer = pl.Trainer(accelerator="hpu", max_epochs=1)
+
+if _HPU_AVAILABLE:
+    # Gaudi HW performs convolution operations with filter (weights) in filters last format
+    from pytorch_lightning.utilities.hpu_device import HPUDeviceUtils
+
+    HPUDeviceUtils.permute_params(model, True)
```
Comment (justusschock) [inline pl_examples/hpu_examples/convolution_example.py:87]: can't we automate this with something like:

```python
if any(isinstance(x, _ConvNd) for x in model.modules())
```

and put this in the accelerator? Usually with PL we try to have code agnostic across hardware and this is likely something to be forgotten

if course, this approach won't work with users invoking functional convolution calls, but the `permute_params` likely won't catch this at all.

File: pytorch_lightning/utilities/hpu_device.py [hunk 3]
```diff
@@ -0,0 +1,58 @@
...
+    # Therefore a re-ordering/permutation of all the convolution weights from KCRS to RSCK format is required
+    # before convolution operations
+    def permute_params(model: "pl.LightningModule", to_filters_last: bool) -> None:
+        """permute the params from filters first (KCRS) to filters last(RSCK) or vice versa.
+
+        and permute from RSCK to KCRS is used for checkpoint saving
+        """
+        with torch.no_grad():
+            for name, param in model.named_parameters():
+                if param.ndim == 4:
```
Comment (justusschock) [inline pytorch_lightning/utilities/hpu_device.py:37]: what about other types of convolutions like 1D or 3D? This won't be catched at all and IMO this is relevant as well.
We cannot assume 2D operations only.

Also we should do some filtering to only apply this to weights of `_ConvNd`. If I have a custom model that for any reason has a 4D parameter that is never used for convolution, I would be pretty confused to have that one permuted like that. This is really error-prone and has to be regarded.

**Comment by rohitgr7** [comment]
can you add some description in the PR or link the issue if there's any? :)

File: pytorch_lightning/utilities/hpu_device.py (hunk 3 above)
Comment (jerome-habana) [reply pytorch_lightning/utilities/hpu_device.py:37]: Agreed @justusschock . We will be removing this need in the subsequent releases. The idea here it to showcase one example to users who might want to work on existing versions of synapse AI

File: docs/source/accelerators/hpu.rst (hunk 1 above)
Comment (jerome-habana) [reply docs/source/accelerators/hpu.rst:119]: This is specifically for the filters. I'll see if I can add some more document for data formats too

**Comment by kaushikb11** [comment]
@jerome-habana let's resolve the conflicts?

**Comment by jerome-habana** [comment]
> @jerome-habana let's resolve the conflicts?

@kaushikb11  I'll close this PR. We are handling this internally now and expect no user script change
